File: solutions/typescript/0092.ts
/**
 * 92. 反转链表 II
 * 一趟扫描：用哑头定位 left 前一个节点 prev，
 * 然后把 prev.next 之后的节点逐个摘下，头插到 prev 之后，
 * 重复 right - left 次即可完成 [left, right] 区间反转。
 *
 * 注：ListNode 类型由 LeetCode 平台注入，本文件用 untyped LNode 接口避免冲突。
 */

interface LNode { val: number; next: LNode | null }

function reverseBetween(head: LNode | null, left: number, right: number): LNode | null {
    const dummy: LNode = { val: 0, next: head };
    let prev: LNode = dummy;
    for (let i = 0; i < left - 1; i++) {
        prev = prev.next as LNode;
    }
    let curr = prev.next as LNode;
    for (let i = 0; i < right - left; i++) {
        const nxt = curr.next as LNode;
        curr.next = nxt.next;
        nxt.next = prev.next;
        prev.next = nxt;
    }
    return dummy.next;
}

export default reverseBetween;

File: solutions/rust/0092.rs
/// 92. 反转链表 II
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     fn new(val: i32) -> Self {
//         ListNode { val, next: None }
//     }
// }

/// Box 链表原地摘取/头插较繁琐，改用与 86 同套路：
/// 提取全部值 → 反转 [left, right) 区间 → 重建链表。
pub fn reverse_between(head: Option<Box<ListNode>>, left: i32, right: i32) -> Option<Box<ListNode>> {
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = node.next;
    }

    let (l, r) = ((left - 1) as usize, (right - 1) as usize);
    vals[l..=r].reverse();

    let mut new_head: Option<Box<ListNode>> = None;
    for &v in vals.iter().rev() {
        new_head = Some(Box::new(ListNode { val: v, next: new_head }));
    }
    new_head
}

impl Solution {
    pub fn reverse_between(head: Option<Box<ListNode>>, left: i32, right: i32) -> Option<Box<ListNode>> {
        reverse_between(head, left, right)
    }
}
